### PR TITLE
Toolbox inventory

### PR DESCRIPTION
**Pull request by chrislovecnm**
This addition to toolbox will allow for the printing of the IoTk ... inventory of all things kops.  
See the list of inventory items in the code comments in the cobra command.  Refactoring has been completed to apply_cluster.go and other code so that code can be re-used.  

This PR will help support hosting all of your own containers, binaries, and other things that are IoTk.

_TODO_

- [x] fix output formatting
- [x] ~~ensure that all of the inventory components are listed~~
- [x] test with existing cluster
- [x] test test and more test

I am not going to make the functional changes in this PR, but I will test to ensure that I have not broken stuff, and move on to another PR that makes functional changes.   I would like to test this before merging.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2419)

### CODE REVIEW

**Comment by justinsb** [comment]
Can you clarify how you expect this command to be used?

**Comment by chrislovecnm** [comment]
This is a diagnostic command to view which inventory items compose a cluster.  For instance, what do I need to put in my repository. We do not have a way to see what is in kops, I can get a full yaml back, but it is missing stuff, like the etcd container.

**Comment by justinsb** [comment]
Looking at the code, I think we need to find a way to do this that doesn't require us to never forget an image.

A few ideas:

* We could output the list of files as we go, by requiring that they all go through a class whose job it was to return a possibly-modified-path.  This way we would have self-hosting by simply doing a dryrun and collecting the list of assets at the end.
* Similarly, look at `fi.HasDependencies` and how dependencies are calculated.  Those are task dependencies, but we could do the same thing for file dependencies: we recursively walk the tasks looking for things that implement `fi.NeedsAssets` (or whatever) and then we collect them all up.
* Once we implement self-hosting, we could have a mock VFS path, and then run the tasks rendering into that mock VFS path.  Then we gather up the dependencies.

The challenge is going to be the nodeup models.  Thankfully nodeup is also declarative.  

So I'd like to see us implement the end-goal here: copying everything to e.g. a private S3 bucket.  And then I'm pretty sure this will fall out as a side-effect.

**Comment by chrislovecnm** [comment]
> Looking at the code, I think we need to find a way to do this that doesn't require us to never forget an image.

Any other ideas on that one?  I am rendering most of them but do have some hard coded :(

> We could output the list of files as we go, by requiring that they all go through a class whose job it was to return a possibly-modified-path. This way we would have self-hosting by simply doing a dryrun and collecting the list of assets at the end.

I am pretty much doing that with the ones that I can.

> Similarly, look at fi.HasDependencies and how dependencies are calculated. Those are task dependencies, but we could do the same thing for file dependencies: we recursively walk the tasks looking for things that implement fi.NeedsAssets (or whatever) and then we collect them all up.

I looked and not sure how that can help us.

> So I'd like to see us implement the end-goal here: copying everything to e.g. a private S3 bucket. And then I'm pretty sure this will fall out as a side-effect.

Yep, that is the end goal.  Let me open an issue for this one.

**Comment by chrislovecnm** [comment]
Added a new issue and updated https://github.com/kubernetes/kops/issues/715

**Comment by chrislovecnm** [comment]
We have a follow up PR that is waiting on this being merged. It is from one of our new contributors.

What is stopping this from being merged?

File: cmd/kops/toolbox_inventory.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+				glog.V(8).Infof("Type of object was %T", v)
+			}
+		}
+	}
+
+	return cluster, ig, nil
+}
+
+// getClientSet returns a clientset.
+func (o *ToolboxInventoryOptions) getClientSet(f *util.Factory) (simple.Clientset, error) {
```
Comment (justinsb) [inline cmd/kops/toolbox_inventory.go]: This function doesn't seem to add anything?

File: protokube/pkg/protokube/etcd_manifest.go [hunk 2]
```diff
@@ -23,6 +23,9 @@ import (
 	"strings"
 )
 
+// TODO: how do I get this out of here or dynamic?
+const ETCD_IMAGE = "gcr.io/google_containers/etcd:2.2.1"
```
Comment (justinsb) [inline protokube/pkg/protokube/etcd_manifest.go]: This isn't actually used?

**Comment by justinsb** [comment]
Did you try making this a target (like dry run)?  I think it could be much simpler if so, but I may be missing something.

**Comment by chrislovecnm** [comment]
Making this a target on create would cause elements to be created. I was attempting to not have cluster components like our state store created. I also had to run apply cloud and ran into issue with ssh keys and other cloud computing components running. I wanted to create something standalone that can be ran before a cluster is created. 

So 

1. Did want to have cloud components created when using the yaml options
2. Ran into problems using the cloudup code in order to generate components like the addons. Code like creating the ssh keys through err's.

File: cmd/kops/toolbox_inventory.go (hunk 1 above)
Comment (chrislovecnm) [inline cmd/kops/toolbox_inventory.go]: Will address

File: protokube/pkg/protokube/etcd_manifest.go (hunk 2 above)
Comment (chrislovecnm) [inline protokube/pkg/protokube/etcd_manifest.go]: Will remove, do you want to keep the etcd version number in protokube?

**Comment by justinsb** [comment]
Dry-run shouldn't create anything.  The SSH thing just sounds like a bug?

Further idea: if we apply your "second" PR first - the one which can redirect images, this might become even easier.  We do a dry-run, and we keep track of what images are queried for redirection.  Those are the images we use.

**Comment by chrislovecnm** [comment]
So it makes sense to put printing the inventory in the create dry run, but we probably need a "kop create assets" for the assets. Thoughts?

Going to look on how to refactor this, but a refactor may be kinda complicated.  I am thinking that we first have a target with "kops create -f mycluster.yaml". That is the use case that I require, and frankly this is an advanced option that yaml users will probably utilize.

Thoughts?

**Comment by robertojrojas** [comment]
My 2 cents. I see the use of the isDryrun variable sprinkled throughout the code. I think that could make it a bit difficult to integrate this code with create.

**Comment by justinsb** [comment]
It's not clear to me what `kops create assets` would do?

I think we merge #2573 and then this work becomes much simpler.  I think it's probably simpler to keep this PR and start a second PR, because I hope this one becomes small.

**Comment by chrislovecnm** [comment]
@justinsb You mentioned a second PR, and I assumed it was the PR that @robertojrojas is working on which is the PR to upload the containers and binaries.  `kop create assets` or whatever we call it would upload the containers.  I think I was confused as you were probably talking about #2573.

I am not sure that #2573 will make this PR much easier.  The challenges is that create --dry-run creates stuff, such as files in s3, and requires an ssh key.  https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/apply_cluster.go#L276 is a perfect example.  I extracted methods in apply_cluster.go and DID not make logic changes.  If you want to use --dry-run target I am going to have to make logic changes and refactor apply_cluster as well.   Maybe my head is over complicating this, but getting the data out of the Tasks was tricky and ended up looking like a unique use case. 

I thought is to have the capability to upload assets as standalone.  Not requiring ssh keys, cluster spec in s3, etc.  Also, the first use case we have for this is with using YAML, so we are adding another PR dependency to this which is https://github.com/kubernetes/kops/pull/2612 😞

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

**Comment by chrislovecnm** [comment]
This is being closed as I need this functionality to test https://github.com/kubernetes/kops/pull/2573, and have included some of this functionality in that PR.

**Pull request closed without merging**